core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,299 @@
+func (f *FwdMgr) Start() error {
+	fwdrs, cnt, err := f.ORM.FindForwardersByChain(utils.Big(*f.evmClient.ChainID()))
+	if err != nil {
+		return errors.Errorf("Failed to retrieve forwarders for chain %d: %s", f.evmClient.ChainID(), err)
+	}
+	f.ctx, f.cancel = context.WithCancel(context.Background())
+	if cnt != 0 {
+		f.initForwardersCache(f.ctx, fwdrs)
+		f.subscribeForwardersLogs(fwdrs)
+	}
+
+	f.authRcvr, err = authorized_receiver.NewAuthorizedReceiver(common.Address{}, f.evmClient)
+	if err != nil {
+		f.logger.Criticalf("Failed to init AuthorizedReceiver")
+	}
+
+	f.offchainAgg, err = offchain_aggregator_wrapper.NewOffchainAggregator(common.Address{}, f.evmClient)
+	if err != nil {
+		f.logger.Criticalf("Failed to init OffchainAggregator")
+	}
+
+	go f.runLoop()
+	return nil
+}
+
+// TODO(samhassan): this should be aware of job type to decide how to fetch senders list.
+func (f *FwdMgr) MaybeForwardTransaction(from common.Address, to common.Address, EncodedPayload []byte) (fwdAddr common.Address, fwdPayload []byte, err error) {
+
+	senders, err := f.getContractSenders(to)
+	if err != nil {
+		return to, EncodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	// Gets current forwarders that are in `to` senders
+	fwdrs, err := f.ORM.FindForwardersInListByChain(utils.Big(*f.evmClient.ChainID()), senders)
+	if err != nil {
+		return to, EncodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	for _, fwdr := range fwdrs {
+		EOAs, err := f.getContractSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnf("Failed to get forwarder senders %s", err)
+			continue
+		}
+		for _, EOA := range EOAs {
+			if EOA == from {
+				forwardedPayload, err := f.getForwardedPayload(to, EncodedPayload)
+				if err != nil {
+					f.logger.Criticalf("Forwarder encoding failed, this should never happen")
+					continue
+				}
+				return fwdr.Address, forwardedPayload, nil
+			}
+		}
+	}
+
+	return to, EncodedPayload, errors.Errorf("Skipping forwarding transaction")
+}
+
+func (f *FwdMgr) getForwardedPayload(dest common.Address, origPayload []byte) ([]byte, error) {
+	callArgs, err := ForwardABI.Inputs.Pack(dest, origPayload)
+	if err != nil {
+		return nil, err
+	}
+
+	dataBytes := append(ForwardABI.ID, callArgs...)
+	return dataBytes, nil
+}
+
+func (f *FwdMgr) getContractSenders(addr common.Address) ([]common.Address, error) {
+	if senders, ok := f.getCachedSenders(addr); ok {
+		return senders, nil
+	}
+	senders, err := f.getAuthorizedSenders(addr)
+	if err != nil {
+		f.logger.Warnf("Failed to call getAuthorizedSenders on %s", addr)
+		return nil, err
+	}
+	f.setCachedSenders(addr, senders)
+	f.subscribeSendersChangedLogs(addr)
+	return senders, nil
+}
+
+func (f *FwdMgr) getAuthorizedSenders(addr common.Address) ([]common.Address, error) {
+	c, err := authorized_receiver.NewAuthorizedReceiverCaller(addr, f.evmClient)
+	if err != nil {
+		f.logger.Errorf("Failed to init forwarder caller: %s", err.Error())
```

Comment:
Again here, we want to log only at the highest level where the error is handled ideally to prevent double logging. Wrap error here instead.